files/galaxy/dynamic_job_rules/production/total_perspective_vortex/roles.yml

```diff
@@ -1,41 +1,62 @@
-            value = not list(ent.tags.filter(tag_value='pulsar')) and not list(entity.tags.filter(tag_value='interactive_pulsar')) # tool has no pulsar tag and so runs on slurm
-        value
+    - id: slurm_training_rule
+      if: |
+        include_tags = []
+        exclude_tags = ['pulsar', 'training-exempt']
+
+        def tag_values_match(entity, match_tag_values=[], exclude_tag_values=[]):
+          return (
+              all([any(entity.tpv_tags.filter(tag_value=tag_value)) for tag_value in match_tag_values])
+              and not any([any(entity.tpv_tags.filter(tag_value=tag_value)) for tag_value in exclude_tag_values])
+          )
+
+        tag_values_match(entity, include_tags, exclude_tags)
+        # helpers.tag_values_match(entity, include_tags, exclude_tags)
       cores: 2
       mem: cores * 3.8  # TODO check multiplier
       context:
         partition: training
       scheduling:
         require:
           - slurm
-    - if: |
-        from tpv.core.entities import Tool
-        value = False
-        for ent in filter(lambda x: isinstance(x, Tool), entities):
-            value = list(ent.tags.filter(tag_value='pulsar')) and not list(ent.tags.filter(tag_value='pulsar-training-large'))
-        value
+          - training
+    - id: pulsar_training_rule
+      if: |
+        include_tags = ['pulsar']
+        exclude_tags = ['pulsar-training-large', 'training-exempt']
+
+        def tag_values_match(entity, match_tag_values=[], exclude_tag_values=[]):
+          return (
+              all([any(entity.tpv_tags.filter(tag_value=tag_value)) for tag_value in match_tag_values])
+              and not any([any(entity.tpv_tags.filter(tag_value=tag_value)) for tag_value in exclude_tag_values])
+          )
+
+        tag_values_match(entity, include_tags, exclude_tags)
+        # helpers.tag_values_match(entity, include_tags, exclude_tags)
       cores: 4
       mem: cores * 2.9  # ratio for pulsar-nci-training
       scheduling:
         require:
           - pulsar  # require pulsar and training: only one place to go
-    - if: |
-        from tpv.core.entities import Tool
-        value = False
-        for ent in filter(lambda x: isinstance(x, Tool), entities):
-            value = list(ent.tags.filter(tag_value='pulsar-training-large'))
-        value
+          - training
+    - id: pulsar_training_large_rule
+      if: |
+        include_tags = ['pulsar-training-large']
+        exclude_tags = ['training-exempt']
+
+        def tag_values_match(entity, match_tag_values=[], exclude_tag_values=[]):
+          return (
+              all([any(entity.tpv_tags.filter(tag_value=tag_value)) for tag_value in match_tag_values])
+              and not any([any(entity.tpv_tags.filter(tag_value=tag_value)) for tag_value in exclude_tag_values])
+          )
+
+        tag_values_match(entity, include_tags, exclude_tags)
+        # helpers.tag_values_match(entity, include_tags, exclude_tags)
```

Comment:
We can make a release of tpv immediately so you can use this instead?

Reply:
That would be good!